sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java

```diff
@@ -169,29 +174,38 @@ public void upload(InputStream data, long length, boolean overwrite) {
     public void uploadWithResponse(InputStream data, long length, ParallelTransferOptions parallelTransferOptions,
         BlobHttpHeaders headers, Map<String, String> metadata, AccessTier tier, BlobRequestConditions requestConditions,
         Duration timeout, Context context) {
+        uploadWithResponse(data, length, new BlobParallelUploadOptions()
+            .setParallelTransferOptions(parallelTransferOptions).setHeaders(headers).setMetadata(metadata).setTier(tier)
+        .setRequestConditions(requestConditions), timeout, context);
+    }
+
+    /**
+     * Creates a new blob, or updates the content of an existing blob.
+     * <p>
+     * To avoid overwriting, pass "*" to {@link BlobRequestConditions#setIfNoneMatch(String)}.
+     *
+     * @param data The data to write to the blob.
+     * @param length The exact length of the data. It is important that this value match precisely the length of the
+     * data provided in the {@link InputStream}.
+     * @param options {@link BlobParallelUploadOptions}
+     * @param timeout An optional timeout value beyond which a {@link RuntimeException} will be raised.
+     * @param context Additional context that is passed through the Http pipeline during the service call.
+     */
+    public Response<BlockBlobItem> uploadWithResponse(InputStream data, long length, BlobParallelUploadOptions options,
+        Duration timeout, Context context) {
+        BlobParallelUploadOptions blobParallelUploadOptions = options == null ? new BlobParallelUploadOptions()
+            : options;
         final ParallelTransferOptions validatedParallelTransferOptions =
-            ModelHelper.populateAndApplyDefaults(parallelTransferOptions);
-        Mono<Object> upload = Mono.fromCallable(() -> {
-            try {
-                // BlobOutputStream will internally handle the decision for single-shot or multi-part upload.
-                BlobOutputStream blobOutputStream = BlobOutputStream.blockBlobOutputStream(client,
-                    validatedParallelTransferOptions, headers, metadata, tier, requestConditions, context);
-                StorageImplUtils.copyToOutputStream(data, length, blobOutputStream);
-                blobOutputStream.close();
-                return null;
-            } catch (IOException e) {
-                Throwable cause = e.getCause();
-                if (cause instanceof BlobStorageException) {
-                    throw logger.logExceptionAsError((BlobStorageException) cause);
-                } else {
-                    throw logger.logExceptionAsError(new UncheckedIOException(e));
-                }
-            }
-            // Subscribing has to happen on a different thread for the timeout to happen properly.
-        }).subscribeOn(Schedulers.elastic());
+            ModelHelper.populateAndApplyDefaults(blobParallelUploadOptions.getParallelTransferOptions());
```

Comment:
Just a thought - Ideally wouldnt this model helper code go in the getter for ParallelTransferOptions in the options bag? or would that be odd for users when they see the getter returns something different from what they set

Reply:
I think it'd be a little odd to manipulate values like that on what otherwise seems to be a simple struct.